core/src/test/scala/unit/kafka/server/AbstractFetcherManagerTest.scala

```diff
@@ -210,4 +217,103 @@ class AbstractFetcherManagerTest {
     verify(fetcher).maybeUpdateTopicIds(Set(tp1), topicIds)
     verify(fetcher).maybeUpdateTopicIds(Set(tp2), topicIds)
   }
+
+  @Test
+  def testExpandThreadPool(): Unit = {
+    val topicPartitions = mockTopicPartition(10, 100)
+    val fetcherManager = new AbstractFetcherManager[AbstractFetcherThread]("fetcher-manager", "fetcher-manager", 10) {
+      override def createFetcherThread(fetcherId: Int, sourceBroker: BrokerEndPoint): AbstractFetcherThread = {
+        new TestResizeFetcherThread(sourceBroker)
+      }
+    }
+    try {
+      resizeAndCheckFetcherPartitionDistribution(fetcherManager, topicPartitions, 25)
+    } finally {
+      fetcherManager.closeAllFetchers()
+    }
+  }
+
+  @Test
+  def testShrinkThreadPool(): Unit = {
+    val topicPartitions = mockTopicPartition(10, 100)
+    val fetcherManager = new AbstractFetcherManager[AbstractFetcherThread]("fetcher-manager", "fetcher-manager", 20) {
+      override def createFetcherThread(fetcherId: Int, sourceBroker: BrokerEndPoint): AbstractFetcherThread = {
+        new TestResizeFetcherThread(sourceBroker)
+      }
+    }
+    try {
+      resizeAndCheckFetcherPartitionDistribution(fetcherManager, topicPartitions, 5)
+    } finally {
+      fetcherManager.closeAllFetchers()
+    }
+  }
+
+  def resizeAndCheckFetcherPartitionDistribution(fetcherManager: AbstractFetcherManager[AbstractFetcherThread], topicPartitions: Set[TopicPartition], fetcherNum: Int): Unit = {
+
+    fetcherManager.addFetcherForPartitions(topicPartitions.map(tp => {
+      val brokerId = Utils.abs(tp.topic().hashCode % 5)
+      val brokerEndPoint = new BrokerEndPoint(brokerId, s"kafka-host-$brokerId", 9092)
```

Comment:
Does using different broker ids bring any value to this test? If it does, it might be good to also verify the broker id when we verify the fetch id.

Reply:
I think each topic partition should be bound with a `BrokerEndPoint` which should its leader. Without leader changed, each topic partition should remain its `BrokerEndPoint` after resizing.  It make sense. 

It seems always satisfied in `addFetcherForPartitions`, but I will add it here in case of `addFetcherForPartitions` changed in the future.